[rendering-modes] New page! (#4667)
withastro/docs: src/content/docs/en/guides/server-side-rendering.mdx

i18nReady: true
---
import PackageManagerTabs from '~/components/tabs/PackageManagerTabs.astro';
import RecipeLinks from '~/components/RecipeLinks.astro';

Astro allows you to choose **on-demand rendering** for some, or all of your pages and endpoints. This is also known as **server-side rendering (SSR)**: generating HTML pages on the server when requested and sending them to the client. An **adapter** is used to run your project on the server and handle these requests.

This on-demand rendering allows you to:
- Implement sessions for login state in your app.
- Render data from an API called dynamically with `fetch()`.
- Deploy your site to a host using an *adapter*.

Consider enabling on-demand server rendering in your Astro project if you need the following:

- **API endpoints**: Create specific pages that function as API endpoints for tasks like database access, authentication, and authorization while keeping sensitive data hidden from the client.

> - **Protected pages**: Handle user access on the server for restrict access to a page based on user privileges.

Author: ```suggestion
- **Protected pages**: Restrict access to a page based on user privileges, by handling user access on the server.
```

export async function GET() {
  return {
    body: JSON.stringify({ message: `This is my static endpoint` }),
  };
}
```

### Opting out of pre-rendering in `hybrid` mode

For a mostly static site configured as `output: hybrid`, add `export const prerender = false` to any files that should be server-rendered on demand:

```js title="src/pages/randomnumber.js" {1}
export const prerender = false;

export async function GET() {
  let number = Math.random();
  return {
    body: JSON.stringify({ number, message: `Here's a random number: ${number}` }),
  };
}
```

### Add an Adapter

To deploy a project in `server` or `hybrid` mode, you also need to add an **adapter**. This is because both of these modes require a server _runtime_: the environment that runs code on the server to generate pages when they are requested. Each adapter allows Astro to output a script that runs your project on a specific runtime, such as Vercel, Netlify or Cloudflare.

You can find both [official and community adapters in our integrations directory](https://astro.build/integrations/?search=&categories%5B%5D=adapters). Choose the one that corresponds to your [deployment environment](/en/guides/deploy/).

> #### `astro add` Install

Reviewer: Should this be capitalized?

```suggestion
#### `astro add` install
```

2. [Add the adapter](/en/reference/configuration-reference/#adapter) to your `astro.config.mjs` file's import and default export, along with your desired `output` mode:

    ```js ins={3,7} {6}
    // astro.config.mjs
    import { defineConfig } from 'astro/config';
    import vercel from '@astrojs/vercel';

    export default defineConfig({
      output: 'server',
      adapter: vercel(),
    });
    ```

    Note that different adapters may also have different configuration settings. Read each adapter's documentation, and apply any necessary config options to your chosen adapter in `astro.config.mjs`

## On-demand rendering features

### HTML streaming

With HTML streaming, a document is broken up into chunks, sent over the network in order, and rendered on the page in that order. In `server` or `hybrid` mode, Astro uses HTML streaming to send each component to the browser as it renders them. This makes sure the user sees your HTML as fast as possible, although network conditions can cause large documents to be downloaded slowly, and waiting for data fetches can block page rendering.

<RecipeLinks slugs={["en/recipes/streaming-improve-page-performance"]}/>

> :::caution[HTML streaming]

Author: ```suggestion
:::caution
```
now that there's a section heading, not needed on the aside itself

> Features that modify the [Response headers](https://developer.mozilla.org/en-US/docs/Glossary/Response_header) are only available at the **page level** when HTML streaming is enabled. (You can't use them inside of components, including layout components.) By the time Astro runs your component code, it has already sent the Response headers and they cannot be modified.

Author: ```suggestion
Features that modify the [Response headers](https://developer.mozilla.org/en-US/docs/Glossary/Response_header) are only available at the **page level**. (You can't use them inside of components, including layout components.) By the time Astro runs your component code, it has already sent the Response headers and they cannot be modified.
```

> Features that modify the [Response headers](https://developer.mozilla.org/en-US/docs/Glossary/Response_header) are only available at the **page level** when HTML streaming is enabled. (You can't use them inside of components, including layout components.) By the time Astro runs your component code, it has already sent the Response headers and they cannot be modified.

Reviewer 2: > when HTML streaming is enabled.

Is this the case. I don't know for sure, but I would be surprised if it is. Usually, the pattern has been that Astro maintains its streaming limitations even during static rendering.

Author: Ah, so my takeaway is it's *correct* but not limited to SSR.

So just confirming we still think that the SSR page is the right place for this. (Even if not limited to SSR, is this a place where people will more likely run into that limitation with what they're trying to do with on-demand rendering?)

> Check your adapter's documentation to see whether HTML streaming is used. It may be the default setting, and may or may not be a configurable option.

Author: ```suggestion
```

---
const cookie = Astro.request.headers.get('cookie');
// ...
---
<html>
  <!-- Page here... -->
</html>
```

#### `Astro.request.method`

The HTTP method used in the request is available as `Astro.request.method`. This works like the browser's [`Request.method`](https://developer.mozilla.org/en-US/docs/Web/API/Request/method). It returns the string representation of the HTTP method used in the request.

```astro title="src/pages/index.astro"
---
console.log(Astro.request.method) // GET (when navigated to in the browser)
---
```

See more details about [`Astro.request`](/en/reference/api-reference/#astrorequest) in the API reference.

### Server Endpoints

A server endpoint, also known as an **API route**, is a special function exported from a `.js` or `.ts` file within the `src/pages/` folder. A powerful feature of server-side rendering on demand, API routes are able to securely execute code on the server.

The function takes an [endpoint context](/en/reference/api-reference/#endpoint-context) and returns a [Response](https://developer.mozilla.org/en-US/docs/Web/API/Response). 

To learn more, see our [Endpoints Guide](/en/core-concepts/endpoints/#server-endpoints-api-routes).

> {/* 
> 
>  */}

Reviewer: I think we can delete this empty comment?

```suggestion
```

Author: Another Sarah Dark Pattern 👻
